### PR TITLE
cockroachdb-first-pr

### PR DESCRIPTION
**Pull request by keith-mcclellan**
initial pr for cockroachdb

## BACKGROUND
* Add information about the background and reason for the change.

Adding CockroachDB helm3 chart to marketplace
-----------------------------------------------------------------------

## Changes
* List the changes that you made

Adding CockroachDB helm3 chart to marketplace
-----------------------------------------------------------------------

## Checklist
- [x] review the [contributing doc](https://github.com/digitalocean/marketplace-kubernetes/blob/master/CONTRIBUTING.md) with steps for both adding or updating your application (if applicable)
------------------------------------------------------------------------

Reviewer: @marketplace-eng

### CODE REVIEW

**Review by pachamaltese — APPROVED** [review]
this looks ok in my opinion
it can be merged

File: stacks/cockroachdb/deploy.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  # use local version of values.yml
+  ROOT_DIR=$(git rev-parse --show-toplevel)
+  values="$ROOT_DIR/stacks/cockroachdb/values.yml"
+else
+  # use github hosted master version of values.yml
+  values="https://raw.githubusercontent.com/digitalocean/marketplace-kubernetes/master/stacks/cockroachdb/values.yml"
+fi
+
+helm upgrade "$STACK" "$CHART" \
+  --atomic \
```
Comment (pachamaltese) [inline stacks/cockroachdb/deploy.sh:29]: remove --atomic

Comment (keith-mcclellan) [reply stacks/cockroachdb/deploy.sh]: done

**Comment by pachamaltese** [comment]
with the last changes I get this error
```
$ kubectl port-forward cockroachdb-0 8080
Error from server (NotFound): pods "cockroachdb-0" not found
```

**Comment by keith-mcclellan** [comment]
can you run `kubectl get pods` please?  I probably just have the pod name wrong.

**Comment by pachamaltese** [comment]
Hi @keith-mcclellan 

Thanks for your contribution!  I've tested locally and I can't get cockroachdb to work :(

This is what I've done on two different kubernetes instances (both fresh), so I'll ping @keladhruv.

# 1: Deploy

```
(base) pacha@pacha-laptop:~cockroachdb$ ./deploy.sh
"cockroachdb" already exists with the same configuration, skipping
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "cockroachdb" chart repository
Update Complete. ⎈Happy Helming!⎈
Release "cockroachdb" does not exist. Installing it now.
NAME: cockroachdb
LAST DEPLOYED: Tue Oct 27 16:18:52 2020
NAMESPACE: cockroachdb
STATUS: deployed
REVISION: 1
NOTES:
CockroachDB can be accessed via port 26257 at the
following DNS name from within your cluster:

cockroachdb-public.cockroachdb.svc.cluster.local

Because CockroachDB supports the PostgreSQL wire protocol, you can connect to
the cluster using any available PostgreSQL client.

Note that because the cluster is running in secure mode, any client application
that you attempt to connect will either need to have a valid client certificate
or a valid username and password.

Finally, to open up the CockroachDB admin UI, you can port-forward from your
local machine into one of the instances in the cluster:

    kubectl port-forward cockroachdb-0 8080

Then you can access the admin UI at https://localhost:8080/ in your web browser.

For more information on using CockroachDB, please see the project's docs at:
https://www.cockroachlabs.com/docs/
```

# 2: Port forward

```
(base) pacha@pacha-laptop:~cockroachdb$ kubectl port-forward cockroachdb-0 8080
Error from server (NotFound): pods "cockroachdb-0" not found
```

# 3: List pods

```
(base) pacha@pacha-laptop:~cockroachdb$ kubectl get pods -A
NAMESPACE     NAME                               READY   STATUS     RESTARTS   AGE
cockroachdb   cockroachdb-0                      0/1     Init:0/1   0          18s
cockroachdb   cockroachdb-1                      0/1     Init:0/1   0          18s
cockroachdb   cockroachdb-2                      0/1     Init:0/1   0          18s
cockroachdb   cockroachdb-init-6g227             0/1     Init:0/1   0          18s
kube-system   cilium-9lmlv                       1/1     Running    0          16m
kube-system   cilium-b2jwz                       1/1     Running    0          16m
kube-system   cilium-operator-6cb976fbcf-mz5r4   1/1     Running    0          18m
kube-system   cilium-operator-6cb976fbcf-zdclz   1/1     Running    0          18m
kube-system   coredns-76bcfddf46-sh9tr           1/1     Running    0          18m
kube-system   coredns-76bcfddf46-wphjv           1/1     Running    0          18m
kube-system   csi-do-node-mgsb9                  2/2     Running    0          16m
kube-system   csi-do-node-v8d6p                  2/2     Running    0          16m
kube-system   do-node-agent-998t4                1/1     Running    0          16m
kube-system   do-node-agent-xv7sp                1/1     Running    0          16m
kube-system   kube-proxy-fzdbq                   1/1     Running    0          16m
kube-system   kube-proxy-sh7t4                   1/1     Running    0          16m
```

**Comment by keith-mcclellan** [comment]
Oh the default config I submitted has TLS enabled, so the CSRs need to be approved.  if you run:

`kubectl get csr`

you should see a list that looks something like:

```
NAME                                    AGE       REQUESTOR                                              CONDITION
default.client.root                     21s       system:serviceaccount:default:my-release-cockroachdb   Pending
default.node.my-release-cockroachdb-0   15s       system:serviceaccount:default:my-release-cockroachdb   Pending
default.node.my-release-cockroachdb-1   16s       system:serviceaccount:default:my-release-cockroachdb   Pending
default.node.my-release-cockroachdb-2   15s       system:serviceaccount:default:my-release-cockroachdb   Pending
...
```

You need to then run `kubectl certificate approve $csr_name` for the CSRs for each pod.  By default there will be 1+${node_count} (the additional one being for the init container, which only runs for about 5 seconds.

I could try to script that in the deploy.sh but that will require that script to run somewhere that kubectl exists and is configured - is that a reasonable assumption?  Alternatively I could have the script cat out the above instructions.

**Comment by pachamaltese** [comment]
@keith-mcclellan @keladhruv 

I was able to complete the suggested steps, but I still cannot get the port forward, it sayd the pod does not exist

```
(base) pacha@pacha-laptop:~/cockroachdb$ kubectl certificate approve cockroachdb.node.cockroachdb-2
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-2 approved
(base) pacha@pacha-laptop:~/cockroachdb$ kubectl get pods -A
NAMESPACE     NAME                               READY   STATUS      RESTARTS   AGE
cockroachdb   cockroachdb-0                      0/1     Running     0          109s
cockroachdb   cockroachdb-1                      0/1     Running     0          109s
cockroachdb   cockroachdb-2                      0/1     Running     0          109s
cockroachdb   cockroachdb-init-qbppx             0/1     Completed   0          109s
kube-system   cilium-hbk44                       1/1     Running     0          4m45s
kube-system   cilium-operator-6cb976fbcf-wxlc4   1/1     Running     0          6m18s
kube-system   cilium-operator-6cb976fbcf-xq562   1/1     Running     0          6m18s
kube-system   cilium-qqzv9                       1/1     Running     0          4m46s
kube-system   coredns-76bcfddf46-qd68g           1/1     Running     0          6m18s
kube-system   coredns-76bcfddf46-s2fr6           1/1     Running     0          6m18s
kube-system   csi-do-node-nb2zg                  2/2     Running     0          4m45s
kube-system   csi-do-node-vvnd2                  2/2     Running     0          4m46s
kube-system   do-node-agent-8cfw5                1/1     Running     0          4m46s
kube-system   do-node-agent-r47hc                1/1     Running     0          4m45s
kube-system   kube-proxy-7jrtg                   1/1     Running     0          4m45s
kube-system   kube-proxy-qtmxg                   1/1     Running     0          4m46s

(base) pacha@pacha-laptop:~/cockroachdb$ kubectl port-forward cockroachdb-0 8080
Error from server (NotFound): pods "cockroachdb-0" not found
```

**Comment by keith-mcclellan** [comment]
can you please do a `kubectl describe pod cockroachdb-0` please?

**Comment by keith-mcclellan** [comment]
I reran locally and my pods started just fine (this was in minikube) - see below.  

```
MacBook-Pro-7:cockroachdb keith$ ./deploy.sh
WARNING: "kubernetes-charts.storage.googleapis.com" is deprecated for "stable" and will be deleted Nov. 13, 2020.
WARNING: You should switch to "https://charts.helm.sh/stable"
"cockroachdb" already exists with the same configuration, skipping
WARNING: "kubernetes-charts.storage.googleapis.com" is deprecated for "stable" and will be deleted Nov. 13, 2020.
WARNING: You should switch to "https://charts.helm.sh/stable"
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "cockroachdb" chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈Happy Helming!⎈
WARNING: "kubernetes-charts.storage.googleapis.com" is deprecated for "stable" and will be deleted Nov. 13, 2020.
WARNING: You should switch to "https://charts.helm.sh/stable"
Release "cockroachdb" does not exist. Installing it now.
NAME: cockroachdb
LAST DEPLOYED: Fri Oct 30 09:51:02 2020
NAMESPACE: cockroachdb
STATUS: deployed
REVISION: 1
NOTES:
CockroachDB can be accessed via port 26257 at the
following DNS name from within your cluster:

cockroachdb-public.cockroachdb.svc.cluster.local

Because CockroachDB supports the PostgreSQL wire protocol, you can connect to
the cluster using any available PostgreSQL client.

Note that because the cluster is running in secure mode, any client application
that you attempt to connect will either need to have a valid client certificate
or a valid username and password.

Finally, to open up the CockroachDB admin UI, you can port-forward from your
local machine into one of the instances in the cluster:

    kubectl port-forward cockroachdb-0 8080

Then you can access the admin UI at https://localhost:8080/ in your web browser.

For more information on using CockroachDB, please see the project's docs at:
https://www.cockroachlabs.com/docs/
MacBook-Pro-7:cockroachdb keith$ kubectl -n cockroachdb get pods
NAME                     READY   STATUS     RESTARTS   AGE
cockroachdb-0            0/1     Init:0/1   0          61s
cockroachdb-1            0/1     Init:0/1   0          60s
cockroachdb-2            0/1     Init:0/1   0          60s
cockroachdb-init-dphkb   0/1     Init:0/1   0          61s
MacBook-Pro-7:cockroachdb keith$ kubectl certificate approve cockroachdb.node.cockroachdb-0
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-0 approved
MacBook-Pro-7:cockroachdb keith$ kubectl certificate approve cockroachdb.node.cockroachdb-1
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-1 approved
MacBook-Pro-7:cockroachdb keith$ kubectl certificate approve cockroachdb.node.cockroachdb-2
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-2 approved
MacBook-Pro-7:cockroachdb keith$ kubectl certificate approve cockroachdb.client.root
certificatesigningrequest.certificates.k8s.io/cockroachdb.client.root approved
MacBook-Pro-7:cockroachdb keith$ kubectl get pods -n cockroachdb
NAME                     READY   STATUS      RESTARTS   AGE
cockroachdb-0            1/1     Running     0          3m43s
cockroachdb-1            1/1     Running     0          3m42s
cockroachdb-2            0/1     Running     0          3m42s
cockroachdb-init-dphkb   0/1     Completed   0          3m43s
MacBook-Pro-7:cockroachdb keith$ kubectl get pods -n cockroachdb
NAME                     READY   STATUS      RESTARTS   AGE
cockroachdb-0            1/1     Running     0          3m45s
cockroachdb-1            1/1     Running     0          3m44s
cockroachdb-2            1/1     Running     0          3m44s
cockroachdb-init-dphkb   0/1     Completed   0          3m45s
MacBook-Pro-7:cockroachdb keith$ kubectl get pods -n cockroachdb
NAME                     READY   STATUS      RESTARTS   AGE
cockroachdb-0            1/1     Running     0          3m47s
cockroachdb-1            1/1     Running     0          3m46s
cockroachdb-2            1/1     Running     0          3m46s
cockroachdb-init-dphkb   0/1     Completed   0          3m47s
```

**Comment by pachamaltese** [comment]
> kubectl describe pod cockroachdb-0

it says
```
$ kubectl describe pod cockroachdb-0
Error from server (NotFound): pods "cockroachdb-0" not found
```
maybe we are missing something about the initialization

**Comment by keith-mcclellan** [comment]
describe is namespaced so you actually need to run:
`kubectl describe pod cockroachdb-0 -n cockroachdb`

**Comment by pachamaltese** [comment]
> describe is namespaced so you actually need to run:
> `kubectl describe pod cockroachdb-0 -n cockroachdb`

thanks, I see these warnings
```
Events:
  Type     Reason                  Age                   From                     Message
  ----     ------                  ----                  ----                     -------
  Warning  FailedScheduling        17m                   default-scheduler        0/3 nodes are available: 3 pod has unbound immediate PersistentVolumeClaims.
  Warning  FailedScheduling        17m                   default-scheduler        0/3 nodes are available: 3 pod has unbound immediate PersistentVolumeClaims.
  Normal   Scheduled               16m                   default-scheduler        Successfully assigned cockroachdb/cockroachdb-0 to pool-0x727dlt5-3szjj
  Normal   SuccessfulAttachVolume  16m                   attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-09994a3b-1802-4491-9ea3-d8343b181c3b"
  Normal   Pulling                 16m                   kubelet                  Pulling image "cockroachdb/cockroach-k8s-request-cert:0.4"
  Normal   Pulled                  16m                   kubelet                  Successfully pulled image "cockroachdb/cockroach-k8s-request-cert:0.4" in 2.225633131s
  Normal   Started                 16m                   kubelet                  Started container init-certs
  Normal   Created                 16m                   kubelet                  Created container init-certs
  Normal   Pulling                 16m                   kubelet                  Pulling image "cockroachdb/cockroach:v20.1.8"
  Normal   Pulled                  16m                   kubelet                  Successfully pulled image "cockroachdb/cockroach:v20.1.8" in 5.539769453s
  Normal   Created                 16m                   kubelet                  Created container db
  Normal   Started                 16m                   kubelet                  Started container db
  Warning  Unhealthy               106s (x175 over 16m)  kubelet                  Readiness probe failed: HTTP probe failed with statuscode: 503
```

**Comment by keith-mcclellan** [comment]
Can you grab the logs from the pod?  command should be something like `kubectl logs cockroachdb-0 -n cockroachdb`

**Comment by pachamaltese** [comment]
> Can you grab the logs from the pod? command should be something like `kubectl logs cockroachdb-0 -n cockroachdb`

sure, please find the log attached
[logs.txt](https://github.com/digitalocean/marketplace-kubernetes/files/5466256/logs.txt)

**Comment by pachamaltese** [comment]
@marketplace-eng now it can be merged!!! after a long review, and thanks to @keith-mcclellan, it works

here are the steps:
```$ bash deploy.sh
"cockroachdb" already exists with the same configuration, skipping
Hang tight while we grab the latest from your chart repositories...
...Successfully got an update from the "cockroachdb" chart repository
Update Complete. ⎈Happy Helming!⎈
Release "cockroachdb" does not exist. Installing it now.
NAME: cockroachdb
LAST DEPLOYED: Fri Oct 30 17:07:19 2020
NAMESPACE: cockroachdb
STATUS: deployed
REVISION: 1
NOTES:
CockroachDB can be accessed via port 26257 at the
following DNS name from within your cluster:

cockroachdb-public.cockroachdb.svc.cluster.local

Because CockroachDB supports the PostgreSQL wire protocol, you can connect to
the cluster using any available PostgreSQL client.

Note that because the cluster is running in secure mode, any client application
that you attempt to connect will either need to have a valid client certificate
or a valid username and password.

Finally, to open up the CockroachDB admin UI, you can port-forward from your
local machine into one of the instances in the cluster:

    kubectl port-forward cockroachdb-0 8080

Then you can access the admin UI at https://localhost:8080/ in your web browser.

For more information on using CockroachDB, please see the project's docs at:
https://www.cockroachlabs.com/docs/
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-2 approved
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-1 approved
certificatesigningrequest.certificates.k8s.io/cockroachdb.node.cockroachdb-0 approved
certificatesigningrequest.certificates.k8s.io/cockroachdb.client.root approved
$ kubectl get pods -n cockroachdb
NAME                     READY   STATUS      RESTARTS   AGE
cockroachdb-0            1/1     Running     0          102s
cockroachdb-1            1/1     Running     0          101s
cockroachdb-2            1/1     Running     0          101s
cockroachdb-init-s9p6l   0/1     Completed   0          101s
$ kubectl port-forward cockroachdb-0 8080
Error from server (NotFound): pods "cockroachdb-0" not found
$ kubectl port-forward cockroachdb-0 8080 -n cockroachdb
Forwarding from 127.0.0.1:8080 -> 8080
Forwarding from [::1]:8080 -> 8080
Handling connection for 8080
Handling connection for 8080
E1030 17:10:49.659118   14404 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:59572: write tcp4 127.0.0.1:8080->127.0.0.1:59572: write: broken pipe
Handling connection for 8080
```

and here's the evidence that it works:
Screenshot from 2020-10-30 17-11-47

**Review by pachamaltese — APPROVED** [review]

**Review by keladhruv — APPROVED** [review]
Based on the review by @pachamaltese , LGTM.

**Comment by jgannondo** [comment]
@keith-mcclellan we ran into a bug with the app after some further testing so we needed to temporarily remove it from the Marketplace. @keladhruv can share more details here but I think a new PR will be required.

**Pull request merged**